### ISSUE TITLE
Simulated Annealing: First mutation always gets accepted?

### ISSUE THREAD

**Issue by Glitchy-Tozier** [argmin-rs/argmin#150]
Hello, I have encountered a scenario which I don't quite understand.

I tried to set the initial temperature (in Simulated annealing) to the smallest possible positive number. (See first line of code-block) According to my current understanding, This should basically result in a greedy hill-climbing algorithm, where only better parameters are accepted.

This isn't fully the case though. The first modification/mutation always seems to get accepted, even though it is way worse.

Code containing the `init_temp`:
```rs
let init_temp = f64::MIN_POSITIVE;

let problem = AnnealingStruct {
    evaluator: Arc::new(evaluator.clone()),
    layout_generator: &pm,
};

// Create new SA solver with some parameters (see docs for details)
// This essentially just prepares the SA solver. It is not run yet, nor does it know anything about the problem it is about to solve.
let solver = SimulatedAnnealing::new(init_temp) // 200.0)
    .unwrap()
    // Optional: Define temperature function (defaults to `SATempFunc::TemperatureFast`)
    .temp_func(SATempFunc::Boltzmann)
    /////////////////////////
    // Stopping criteria   //
    /////////////////////////
    // Optional: stop if there was no accepted solution after 1000 iterations
    .stall_accepted(params.stall_accepted);

// Create and run the executor, which will apply the solver to the problem, given a starting point (`init_param`)
let res = Executor::new(problem, solver, init_layout)
    // Optional: Attach a observer
    .add_observer(ArgminSlogLogger::term(), ObserverMode::Always) //Every(100))
    // Optional: Set maximum number of iterations (defaults to `std::u64::MAX`)
    .max_iters(params.max_iters)
    // Optional: Set target cost function value (defaults to `std::f64::NEG_INFINITY`)
    .run()
    .unwrap();
```

Log:
```rs
Nov 28 16:53:58.373 INFO Simulated Annealing, max_iters: 10000, initial_temperature: 0.000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000022250738585072014, stall_iter_accepted_limit: 1000, stall_iter_best_limit: 18446744073709551615, reanneal_fixed: 18446744073709551615, reanneal_accepted: 18446744073709551615, reanneal_best: 18446744073709551615

Nov 28 16:53:58.528 INFO iter: 0, cost: 269.73266205959345, best_cost: 269.73266205959345, cost_func_count: 1, grad_func_count: 0, jacobian_func_count: 0, hessian_func_count: 0, modify_func_count: 1, t: 0.000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000032101030212800104, new_be: true, acc: true, st_i_be: 0, st_i_ac: 0, ra_i_fi: 1, ra_i_be: 0, ra_i_ac: 0, ra_fi: false, ra_be: false, ra_ac: false, time: 0.154631085
Nov 28 16:53:58.677 INFO iter: 1, cost: 579.427167261033, best_cost: 269.73266205959345, cost_func_count: 2, grad_func_count: 0, jacobian_func_count: 0, hessian_func_count: 0, modify_func_count: 2, t: 0.00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000002025349508155188, new_be: false, acc: true, st_i_be: 1, st_i_ac: 0, ra_i_fi: 2, ra_i_be: 1, ra_i_ac: 0, ra_fi: false, ra_be: false, ra_ac: false, time: 0.14926356300000002
Nov 28 16:53:58.827 INFO iter: 2, cost: 579.427167261033, best_cost: 269.73266205959345, cost_func_count: 3, grad_func_count: 0, jacobian_func_count: 0, hessian_func_count: 0, modify_func_count: 3, t: 0.00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000001605051510640005, new_be: false, acc: false, st_i_be: 2, st_i_ac: 1, ra_i_fi: 3, ra_i_be: 2, ra_i_ac: 1, ra_fi: false, ra_be: false, ra_ac: false, time: 0.150058625
Nov 28 16:53:58.977 INFO iter: 3, cost: 576.9572329862979, best_cost: 269.73266205959345, cost_func_count: 4, grad_func_count: 0, jacobian_func_count: 0, hessian_func_count: 0, modify_func_count: 4, t: 0.00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000001382516120265875, new_be: false, acc: true, st_i_be: 3, st_i_ac: 0, ra_i_fi: 4, ra_i_be: 3, ra_i_ac: 0, ra_fi: false, ra_be: false, ra_ac: false, time: 0.149817584
.
.
.
Nov 28 17:12:54.538 INFO iter: 7334, cost: 411.77879579661055, best_cost: 269.73266205959345, cost_func_count: 7335, grad_func_count: 0, jacobian_func_count: 0, hessian_func_count: 0, modify_func_count: 7335, t: 0.000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000002499928774082615, new_be: false, acc: false, st_i_be: 7334, st_i_ac: 5, ra_i_fi: 7335, ra_i_be: 7334, ra_i_ac: 5, ra_fi: false, ra_be: false, ra_ac: false, time: 0.157413196
Nov 28 17:12:54.694 INFO iter: 7335, cost: 411.77879579661055, best_cost: 269.73266205959345, cost_func_count: 7336, grad_func_count: 0, jacobian_func_count: 0, hessian_func_count: 0, modify_func_count: 7336, t: 0.000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000002499890490262483, new_be: false, acc: false, st_i_be: 7335, st_i_ac: 6, ra_i_fi: 7336, ra_i_be: 7335, ra_i_ac: 6, ra_fi: false, ra_be: false, ra_ac: false, time: 0.15598256600000002
Nov 28 17:12:54.851 INFO iter: 7336, cost: 411.77879579661055, best_cost: 269.73266205959345, cost_func_count: 7337, grad_func_count: 0, jacobian_func_count: 0, hessian_func_count: 0, modify_func_count: 7337, t: 0.000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000002499852212832205, new_be: false, acc: true, st_i_be: 7336, st_i_ac: 0, ra_i_fi: 7337, ra_i_be: 7336, ra_i_ac: 0, ra_fi: false, ra_be: false, ra_ac: false, time: 0.157277205
Nov 28 17:12:55.007 INFO iter: 7337, cost: 411.77879579661055, best_cost: 269.73266205959345, cost_func_count: 7338, grad_func_count: 0, jacobian_func_count: 0, hessian_func_count: 0, modify_func_count: 7338, t: 0.00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000249981394178982, new_be: false, acc: false, st_i_be: 7337, st_i_ac: 1, ra_i_fi: 7338, ra_i_be: 7337, ra_i_ac: 1, ra_fi: false, ra_be: false, ra_ac: false, time: 0.15549885400000002

```

**Comment by Glitchy-Tozier** [comment]
In case you're interested, the the initial cost is `257.44832199665365`.

Also, this weird behaviour doesn't seem to stem from the scary-looking, enormous `f64::MIN_POSITIVE`. If I replace it with `0.0000000000001`, we have the same problem on our hands.

**Comment by Glitchy-Tozier** [comment]
Another weird but probably unrelated occurrence is that sometimes, even when we're using `.add_observer(ArgminSlogLogger::term(), ObserverMode::NewBest)`, new lines are added even though the cost of the iterations stays perfectly consistent (which also seems weird to me). This happens maybe 1/4th of the time.

Should I open a new issue regarding this problem?



```rs
// Create and run the executor, which will apply the solver to the problem, given a starting point (`init_param`)
let res = Executor::new(problem, solver, init_layout)
    // Optional: Attach a observer
    .add_observer(ArgminSlogLogger::term(), ObserverMode::NewBest) //Always) //Every(100))
    // Optional: Set maximum number of iterations (defaults to `std::u64::MAX`)
    .max_iters(params.max_iters)
    .run()
    .unwrap();
```

```rs
Nov 28 17:42:49.124 INFO Simulated Annealing, max_iters: 10000, initial_temperature: 0.000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000022250738585072014, stall_iter_accepted_limit: 1000, stall_iter_best_limit: 18446744073709551615, reanneal_fixed: 18446744073709551615, reanneal_accepted: 18446744073709551615, reanneal_best: 18446744073709551615
Nov 28 17:42:49.278 INFO iter: 0, cost: 258.97784150324475, best_cost: 258.97784150324475, cost_func_count: 1, grad_func_count: 0, jacobian_func_count: 0, hessian_func_count: 0, modify_func_count: 1, t: 0.000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000032101030212800104, new_be: true, acc: true, st_i_be: 0, st_i_ac: 0, ra_i_fi: 1, ra_i_be: 0, ra_i_ac: 0, ra_fi: false, ra_be: false, ra_ac: false, time: 0.154570444
Nov 28 17:42:50.023 INFO iter: 5, cost: 258.97784150324475, best_cost: 258.97784150324475, cost_func_count: 6, grad_func_count: 0, jacobian_func_count: 0, hessian_func_count: 0, modify_func_count: 6, t: 0.00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000001143461767537116, new_be: true, acc: true, st_i_be: 0, st_i_ac: 0, ra_i_fi: 6, ra_i_be: 0, ra_i_ac: 0, ra_fi: false, ra_be: false, ra_ac: false, time: 0.14868594200000002
Nov 28 17:42:50.174 INFO iter: 6, cost: 258.97784150324475, best_cost: 258.97784150324475, cost_func_count: 7, grad_func_count: 0, jacobian_func_count: 0, hessian_func_count: 0, modify_func_count: 7, t: 0.000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000010700343404266705, new_be: false, acc: false, st_i_be: 1, st_i_ac: 1, ra_i_fi: 7, ra_i_be: 1, ra_i_ac: 1, ra_fi: false, ra_be: false, ra_ac: false, time: 0.15117486800000002
Nov 28 17:42:50.328 INFO iter: 7, cost: 258.97784150324475, best_cost: 258.97784150324475, cost_func_count: 8, grad_func_count: 0, jacobian_func_count: 0, hessian_func_count: 0, modify_func_count: 8, t: 0.00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000001012674754077594, new_be: false, acc: false, st_i_be: 2, st_i_ac: 2, ra_i_fi: 8, ra_i_be: 2, ra_i_ac: 2, ra_fi: false, ra_be: false, ra_ac: false, time: 0.15384032
Nov 28 17:42:50.481 INFO iter: 8, cost: 258.97784150324475, best_cost: 258.97784150324475, cost_func_count: 9, grad_func_count: 0, jacobian_func_count: 0, hessian_func_count: 0, modify_func_count: 9, t: 0.00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000966337298576854, new_be: false, acc: false, st_i_be: 3, st_i_ac: 3, ra_i_fi: 9, ra_i_be: 3, ra_i_ac: 3, ra_fi: false, ra_be: false, ra_ac: false, time: 0.153119312
Nov 28 17:42:50.634 INFO iter: 9, cost: 258.97784150324475, best_cost: 258.97784150324475, cost_func_count: 10, grad_func_count: 0, jacobian_func_count: 0, hessian_func_count: 0, modify_func_count: 10, t: 0.000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000009279278723088335, new_be: false, acc: false, st_i_be: 4, st_i_ac: 4, ra_i_fi: 10, ra_i_be: 4, ra_i_ac: 4, ra_fi: false, ra_be: false, ra_ac: false, time: 0.152551731
```

**Comment by stefan-k** [comment]
Thanks for reporting this.

I'm unsure if this is a bug. I had a quick look at the code but couldn't find anything that could cause this, but I will have a closer look in the next couple of days. 
However, my guess is that your temperature is way too small. If I counted the zeros correctly, your temperature is 10^-13. Since `new_cost` is larger than `prev_cost`, acceptance is calculated with `1 / (1 + exp((next_cost - prev_cost) / current_temperature))`. Given your temperature, the exponential of a very large number is calculated which may cause floating point problems. Interestingly though it does seem to work after the first iteration. 
Either way, this is something that the SA implementation should handle properly. However, for your case you might be able to work around this by choosing a larger yet still small temperature for now. 

> I tried to set the initial temperature (in Simulated annealing) to the smallest possible positive number. (See first line of code-block) According to my current understanding, This should basically result in a greedy hill-climbing algorithm, where only better parameters are accepted.

Side note: This is only true if you ignore the temperature inside your `modify` function. If you follow the example implementation then your modifications will also be rather small, therefore you won't be moving far in the parameter space (or even not at all).

> Another weird but probably unrelated occurrence is that sometimes, even when we're using .add_observer(ArgminSlogLogger::term(), ObserverMode::NewBest), new lines are added even though the cost of the iterations stays perfectly consistent (which also seems weird to me). This happens maybe 1/4th of the time.

I assume this is likely due to differences in cost, which are beyond the digits that are printed to the screen. If your changes in the `modify` function are scaled with your small temperature, your movements will be small and therefore also the differences in cost are likely to be very small.

**Comment by Glitchy-Tozier** [comment]
> However, my guess is that your temperature is way too small. If I counted the zeros correctly, your temperature is 10^-13. Since new_cost is larger than prev_cost, acceptance is calculated with 1 / (1 + exp((next_cost - prev_cost) / current_temperature)). Given your temperature, the exponential of a very large number is calculated which may cause floating point problems. Interestingly though it does seem to work after the first iteration.
Either way, this is something that the SA implementation should handle properly. However, for your case you might be able to work around this by choosing a larger yet still small temperature for now.

That is a great idea, unfortunately it doesn't seem to work. I tried `0.01`, but that too always gets accepted at the first iteration.

> Side note: This is only true if you ignore the temperature inside your `modify` function. If you follow the example implementation then your modifications will also be rather small, therefore you won't be moving far in the parameter space (or even not at all).

I'm not 100% sure what you mean. I performed the smallest possible modifications. The starting cost is **240**, which already is pretty optimized. Usually the first iterations have a cost of **~330**, but with really bad luck, we might end up at **500**.
The best possible solution found so far has a cost of **208**.

> I assume this is likely due to differences in cost, which are beyond the digits that are printed to the screen. If your changes in the `modify` function are scaled with your small temperature, your movements will be small and therefore also the differences in cost are likely to be very small.

Let's move this to #151, I have found some more weird stuff :see_no_evil:

**Comment by Glitchy-Tozier** [comment]
By the way, if you ever want to look at the full code, you can find it [here](https://github.com/dariogoetz/keyboard_layout_optimizer/pull/6).
Or just ask me, and I'll post the code-snippet. :)

**Comment by Glitchy-Tozier** [comment]
Another thing that might be worth mentioning: The temperature in the first log seems to **always be larger** than the initial temperature. For example, with `let init_temp = 0.01`, I get this log:

```rs
Nov 28 21:17:18.569 INFO Simulated Annealing, max_iters: 10000, initial_temperature: 0.01, stall_iter_accepted_limit: 1000, stall_iter_best_limit: 18446744073709551615, reanneal_fixed: 18446744073709551615, reanneal_accepted: 18446744073709551615, reanneal_best: 18446744073709551615

println from inside `apply()`: Cost: 302.38946704551745
Nov 28 21:17:18.730 INFO iter: 0, cost: 302.38946704551745, best_cost: 302.38946704551745, cost_func_count: 1, grad_func_count: 0, jacobian_func_count: 0, hessian_func_count: 0, modify_func_count: 1, t: 0.014426950408889635, new_be: true, acc: true, st_i_be: 0, st_i_ac: 0, ra_i_fi: 1, ra_i_be: 0, ra_i_ac: 0, ra_fi: false, ra_be: false, ra_ac: false, time: 0.16084857200000002
```

**Comment by stefan-k** [comment]
> Another thing that might be worth mentioning: The temperature in the first log seems to **always be larger** than the initial temperature.

You are using `SATempFunc::Boltzmann` which adapts the temperature according to `t_init / ln(i)` where `i` is the iteration number. Since `ln(1) = 0`, I'm using `t_init / ln(i + 1)` to avoid dividing by `0`, therefore for the first iteration `t_init / ln(1 + 1) = 0.01 / 0.69314718056 = 0.0144269504`. 

> I'm not 100% sure what you mean.

Nevermind, seems like you're ignoring the temperature in `modify`. 

> That is a great idea, unfortunately it doesn't seem to work. I tried 0.01, but that too always gets accepted at the first iteration.

Thanks for trying, I will have a look in the next days.

**Comment by stefan-k** [comment]
I believe the main issue mentioned here was fixed in #153, therefore I will close it. Feel free to reopen if you have further questions regarding this issue. Thanks for reporting!
